include/pybind11/pybind11.h

```diff
@@ -150,8 +157,20 @@ class cpp_function : public function {
 #if defined(__GNUG__) && !defined(__clang__) && __GNUC__ >= 6
 #  pragma GCC diagnostic pop
 #endif
+#if defined(__GNUG__) && !PYBIND11_HAS_STD_LAUNDER
+#  pragma GCC diagnostic push
+#  pragma GCC diagnostic ignored "-Wstrict-aliasing"
+#endif
+            // UB without std::launder, but without breaking ABI and/or
+            // a significant refactoring it's "impossible" to solve'
             if (!std::is_trivially_destructible<Func>::value)
-                rec->free_data = [](function_record *r) { ((capture *) &r->data)->~capture(); };
+                rec->free_data = [](function_record *r) {
+                    auto cap = PYBIND11_STD_LAUNDER((capture *) &r->data);
+                    cap->~capture();
```

Comment:
MSVC tells me that I'm not using `cap`. If I put it all on one line MSVC says that I am discarding the return value of `launder`, which MSVC decided to mark as `[[nodiscard]]`. Note that this affects MSVC 2015 and MSVC latest, but not MSVC 2017...